### PR TITLE
1 create search api prototype

### PR DESCRIPTION
**Pull request by InnaVays**
Questions, Answers, Refs

### CODE REVIEW

**Comment by iliatimofeev** [comment]
Redis provides a high-level API to manipulate with [Pydantic](https://pypi.org/project/pydantic/) (which are native to FastAPI) objects, take a look into [blog post](https://redis.com/blog/introducing-redis-om-for-python/),  [documentation](https://redis.io/docs/stack/get-started/tutorials/stack-python/) and [source code](https://github.com/redis/redis-om-python) .  Additionally there are some external packages that do it like [pydantic-redis](https://sopherapps.github.io/pydantic-redis/explanation/why-use-orms/). 

- [ ] Please explore that opportunity and consider removing low level querying and encoding in your code.  I will reference this suggesion  as **redis_om**.

Please refactor your code into separate modules and classes. `utils.py` intends to be a module for small system utilities, not for putting all your notebook cells into it. `deps.py` is for dependency injection into FastAPI request processing, not for heavy lifting. API implementation files are for implementing API calls; all the heavy lifting should be done in separate modules.


Please refactor your code into separate modules and classes. `utils.py` intends to be a module for small system utilities, not for putting all your notebook cells into it. `deps.py` is for dependency injection into FastAPI request processing, not for heavy lifting. API implementation files are for implementing API calls; all the heavy lifting should be done in separate modules. You have different entities; they should be implemented as classes. Some highly connected entities could be in one module (file), but by default, each entity has one module.

- [ ] `document.py` is for Document class. A document can represent itself as a list (or iterator) of DocumentBlocks. You get a document from the API and just ask it for a DocumentBlocks providing the expected size of the block . 

- [ ] `collection.py` is for `Collection` class that could store and search `DocumentBlock` all the internals of Redis, Cohere, and how exactly you do it. Just a simple interface: add, update, delete, list, search. So when you are implementing an API call, you know nothing about embeddings, or the Redis interface, you just get a list of DocumentBlocks from the document and add them to the collection. Or just get your text query and ask collection for DocumentBlocks

- [ ] `llm.py` currently put functions on question generation and answering on a set of DocumentBlocks into this module.

Test. 

- [ ] Tests **must** be implemented.

File: app/api/api_v1/endpoints/document.py [hunk 1]
```diff
@@ -1,85 +1,123 @@
...
 
 @router.get("/{collection_name}/", response_model=List[str])
-def list_documents(collection_name: str) -> Any:
-    """
-    Get a list of document IDs in the collection.
-    """
-    collection = get_collection_by_name(collection_name)
-    return list(collection.documents.keys())
+def list_documents(collection:  deps.CollectionDep) -> Any:
+    keys = [key for key in collection.connection.scan_iter(f"{collection.name}*")]
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:18]: Collection class

File: app/api/api_v1/endpoints/document.py [hunk 2]
```diff
@@ -1,85 +1,123 @@
...
+@router.post("/{collection_name}/", response_model=str)
+def add_documents( collection: deps.CollectionDep, documents: List[schemas.Document]) -> Any:
     """
     Dump LIST of new documents to a collection.
     """
-    collection = get_collection_by_name(collection_name)
- 
-    # Validate and add each document to the collection
+    # instantiate and fill redis pipeline
+    pipe = deps.db.connection.pipeline()
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:39]: Collection class

File: app/api/api_v1/endpoints/document.py [hunk 3]
```diff
@@ -1,85 +1,123 @@
...
-                                                                localeCode = document.localeCode,
-                                                                title = document.title,
-                                                                description = document.description,
-                                                                render = document.render)
-    return list(collection.documents.keys())
-
-@router.get("/{collection_name}/{path}", response_model=schemas.Document)
-def read_document(collection_name: str, path: str) -> Any:
+        # define key
+        blocks = utils.extract_info_blocks(document.render, threshold=utils.THRESHOLD)
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:42]: Document class

File: app/api/api_v1/endpoints/document.py [hunk 4]
```diff
@@ -1,85 +1,123 @@
...
+        encoded_blocks = [ utils.encode_blocks(deps.db.embedding,block['text']) for block in blocks ] 
+        for i,vector in enumerate(encoded_blocks):
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:44]: Collection class

File: app/api/api_v1/endpoints/document.py [hunk 5]
```diff
@@ -1,85 +1,123 @@
...
+                                 "render_id": blocks[i]['render_id'], 
+                                 "chunk_num": blocks[i]['chunk_num'],
+                                 "text": blocks[i]['text'],
+                                 "content": blocks[i]['content'],
+                                 "locale": document.localeCode }
+
+            key = collection.get_a_key(document.path, i)
+            
+            document_metadata[utils.VECTOR_FIELD_NAME] = np.array(vector).astype(np.float32).tobytes()
+            # HSET
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:57]: redis_om

File: app/api/api_v1/endpoints/document.py [hunk 6]
```diff
@@ -1,85 +1,123 @@
...
-    if path in collection.documents:
-        collection.documents[path].localeCode = document.localeCode
-        collection.documents[path].title = document.title
-        collection.documents[path].description = document.description
-        collection.documents[path].render = document.render
-        return collection.documents[path]
+    if len(keys_to_delete) < 1:
+        raise HTTPException(status_code=404, detail="Document not found")
+
+    for key in keys_to_delete:
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:86]: Collection class

File: app/api/api_v1/endpoints/document.py [hunk 7]
```diff
@@ -1,85 +1,123 @@
...
+        deps.db.connection.delete(key)
+
+    # instantiate and fill redis pipeline
+    pipe = deps.db.connection.pipeline()
+    blocks = utils.extract_info_blocks(document.render, threshold = utils.THRESHOLD)
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:91]: Document class return list of DocumentBlocks

File: app/api/api_v1/endpoints/document.py [hunk 8]
```diff
@@ -1,85 +1,123 @@
...
+    encoded_blocks = [ utils.encode_blocks(deps.db.embedding,block['text']) for block in blocks ] 
+    for i,vector in enumerate(encoded_blocks):
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:93]: Collection class accepts DocumentBlocks

File: app/api/api_v1/endpoints/document.py [hunk 9]
```diff
@@ -1,85 +1,123 @@
...
-        del collection.documents[path]
-        return f"Document {path} deleted"
+    keys_to_delete = collection.search_by_path(path)
+    if len(keys_to_delete) < 1:
+        raise HTTPException(status_code=404, detail="Document not found")
 
-    raise HTTPException(status_code=404, detail="Document not found")
\ No newline at end of file
+    for key in keys_to_delete:
+        deps.db.connection.delete(key)
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/document.py:121]: Collection

File: app/api/api_v1/endpoints/search.py [hunk 10]
```diff
@@ -1,39 +1,81 @@
...
-                )
-                results.append(result)
-
-    return schemas.SearchResultFull(
-        results=results,
-        suggestions=suggestions,
-        totalHits=len(results) + len(suggestions)
-    )
+    for question in questions:
+        query_embedding = utils.encode_blocks(deps.db.embedding, question)
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/search.py:33]: Collection should accept plain text query

File: app/api/api_v1/endpoints/search.py [hunk 11]
```diff
@@ -1,39 +1,81 @@
...
-from app.shared_state import existing_collections
+from app.core.config import settings  # pylint: disable=C0415
 
 router = APIRouter()
 
+import lxml.html
+import numpy as np
+
+
 @router.get("/", response_model = schemas.SearchResultFull)
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/search.py:16]:
```python
@router.get("/{collection_name} ...."
```

I missed that previously.
We should search in the exact collection. Each deployment of Joan will have a unique collection. There is no reason to return documents from other collections, as for now at least.

File: app/api/api_v1/endpoints/search.py [hunk 12]
```diff
@@ -1,39 +1,81 @@
...
-        results=results,
-        suggestions=suggestions,
-        totalHits=len(results) + len(suggestions)
-    )
+    for question in questions:
+        query_embedding = utils.encode_blocks(deps.db.embedding, question)
+        # A list to store the search results for the current question from all collections
+        for collection in existing_collections:
+            #print(f"Search in {collection} ")
+            redis_search_results = index.search_vectors_knn(redis_connect = deps.db.connection,
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/search.py:37]: Redis_om
Consider to wrap `index.search_vectors_knn` with method that returns list of results Pydantic objects. I guess you can find in the source code of  [redis_om](https://github.com/redis/redis-om-python/blob/87d3a080adf6e9dc2bda1b38204b5b341876f16e/aredis_om/model/model.py#L1371)  examples on how to do it properly.

File: app/api/api_v1/endpoints/search.py [hunk 13]
```diff
@@ -1,39 +1,81 @@
...
+                #print(f"Add result from {collection}")
+
+        # Sort the search results for the current question based on the scores and select top_k results
+        sorted_list = sorted(question_results, key=lambda x: float(x['dist']))
+        top_results[question] = sorted_list[:utils.TOP_LINKS]
+        #print(' Store best 5 for a question!')
 
-    
\ No newline at end of file
+        for redis_doc in top_results[question]:
```
Comment (iliatimofeev) [inline app/api/api_v1/endpoints/search.py:52]: Please explore `redis_om` and consider to replace that unreadable `redis_path = redis_dict['path'.encode('utf-8')].decode('utf-8') …` into `block = DocumentBlock.get( result.id)`  and you’ll get access to `block.path` using [redis_om](https://github.com/redis/redis-om-python).

File: app/api/deps.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        # The parameters here are placeholders, replace with your actual parameters
+        return Redis( host = settings.REDIS_HOST,
+                    port = settings.REDIS_PORT,
+                    password = settings.REDIS_PASSWORD)
+    @cached_property
+    def embedding(self):
+        # Initialize the connection to Cohere
+        return cohere.Client(settings.COHERE_API_KEY)
+      
+class CollectionDB():
```
Comment (iliatimofeev) [inline app/api/deps.py:26]: Move to separete file `collection.py`

File: app/api/index.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
+
+
+from redis.commands.search.field import TagField, VectorField
+from redis.commands.search.indexDefinition import IndexDefinition, IndexType
+from redis.commands.search.query import Query
+import numpy as np
+
+def create_index(redis_connect, index_name, doc_prefix, vector_name, vector_dimensions, distance_metric='IP'):
```
Comment (iliatimofeev) [inline app/api/index.py:8]: collection.py

File: app/api/utils.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+N_QUESTIONS = 3
+
+
+# Load your OpenAI API key from an environment variable or secret management service
+api_key = 
+
+# Authenticate with the API
+openai.api_key = api_key
+
+def split_text(text, threshold):
```
Comment (iliatimofeev) [inline app/api/utils.py:27]: document.py

File: app/api/utils.py [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    return result
+
+# Function to encode blocks using Cohere
+def encode_blocks(cohere_client,text, cohere_model = COHERE_MODEL):
+    """
+    Function to encode blocks using Cohere
+    """
+    return np.array(cohere_client.embed(texts=[text],  model=cohere_model).embeddings[0]).astype(np.float32)
+
+def generate_questions(prompt, model = "gpt-3.5-turbo", n = 5):
```
Comment (iliatimofeev) [inline app/api/utils.py:92]: llm.py

**Review by iliatimofeev — CHANGES_REQUESTED** [review]

File: tests/test_api.py [hunk 18]
```diff
@@ -3,39 +3,32 @@
...
 
 from starlette.testclient import TestClient
 import json
 from typing import List
 
 from app.api.api_v1.api import collection, document, search
-from app.schemas import Collection, CollectionCreate, Document, DocumentUpdate, SearchResultFull, SearchResult
-from app.shared_state import existing_collections
+from app.schemas import Collection, Document
+from app.api import utils
```
Comment (iliatimofeev) [inline tests/test_api.py:14]: https://docs.pytest.org/en/6.2.x/fixture.html

File: tests/test_api.py [hunk 19]
```diff
@@ -110,32 +94,49 @@ def test_delete_collection():
...
-    documents = [
-        Document(path="test_document_1", localeCode="en", title="Test Document 1", description="Test Document 1",
-                 render="test document 1 render").dict(),
-        Document(path="test_document_2", localeCode="en", title="Test Document 2", description="Test Document 2",
-                 render="test document 2 render").dict()
-    ]
-    response = client.post("api/v1/documents/test_collection/", json = documents)
+    new_documents = [ test_doc ]
+    response = client.post("api/v1/documents/test_collection/", json = new_documents)
     assert response.status_code == 200
```
Comment (iliatimofeev) [inline tests/test_api.py:106]: Move lxml tests into unit/test_document.py

**Review by iliatimofeev — CHANGES_REQUESTED** [review]

File: tests/test_api.py (hunk 18 above)
Comment (InnaVays) [reply tests/test_api.py:14]: Done

File: app/api/deps.py (hunk 14 above)
Comment (InnaVays) [reply app/api/deps.py:26]: done

File: app/schemas/search.py [hunk 20]
```diff
@@ -10,5 +10,7 @@ class SearchResult(BaseModel):
 
 class SearchResultFull(BaseModel):
```
[the remark this answers is not in the document]
Comment (InnaVays) [reply app/schemas/search.py:11]: done

File: app/api/api_v1/endpoints/document.py (hunk 2 above)
Comment (InnaVays) [reply app/api/api_v1/endpoints/document.py:39]: done

File: app/api/api_v1/endpoints/document.py (hunk 5 above)
Comment (InnaVays) [reply app/api/api_v1/endpoints/document.py:57]: done

File: app/api/api_v1/endpoints/document.py (hunk 9 above)
Comment (InnaVays) [reply app/api/api_v1/endpoints/document.py:121]: done

File: app/api/index.py (hunk 15 above)
Comment (InnaVays) [reply app/api/index.py:8]: done

File: app/api/api_v1/endpoints/document.py (hunk 3 above)
Comment (InnaVays) [reply app/api/api_v1/endpoints/document.py:42]: done

File: app/api/api_v1/endpoints/document.py (hunk 4 above)
Comment (InnaVays) [reply app/api/api_v1/endpoints/document.py:44]: done

File: tests/test_api.py (hunk 19 above)
Comment (InnaVays) [reply tests/test_api.py:106]: done

File: app/api/api_v1/endpoints/document.py (hunk 8 above)
Comment (InnaVays) [reply app/api/api_v1/endpoints/document.py:93]: done

**Comment by iliatimofeev** [comment]
1. Redis runs locally in the codespaces container.
2. Cohere and OpenaAI keys are available in codespace
3. Rebased to main
4. Github runs tests on PRs for main. Test results are visible in comments.
5. fixed typo in test_document, and it fails now.

File: tests/test_api.py [hunk 21]
```diff
@@ -5,137 +5,137 @@
...
 from typing import List
 
 from app.api.api_v1.api import collection, document, search
-from app.schemas import Collection, CollectionCreate, Document, DocumentUpdate, SearchResultFull, SearchResult
-from app.shared_state import existing_collections
+from app.schemas import Collection, Document
+from app.api import llm
+import json
 
 client = TestClient(app)
```
Comment (iliatimofeev) [inline tests/test_api.py:17]: That should be fixure too

**Review by iliatimofeev — APPROVED** [review]

**Pull request merged**